### PR TITLE
workaround mangle collision between decltype(Rf_error) and decltype(Rf_warning)

### PR DESCRIPTION
**Pull request by bkietz**
I also cleaned up the signatures of detail:: functions, hopefully it's more readable

### CODE REVIEW

**Comment by romainfrancois** [comment]
I'm still getting the same problem on my current branch of arrow: 

```
==> R CMD INSTALL --preclean --no-multiarch --with-keep.source r

* installing to library ‘/Users/romainfrancois/.R/library/4.0’
*** Generating code with data-raw/codegen.R
* installing *source* package ‘arrow’ ...
** using staged installation
Warning messages:
1: package ‘decor’ was built under R version 4.0.2 
2: package ‘vctrs’ was built under R version 4.0.2 
*** > 378 functions decorated with [[arrow|s3::export]]
*** > generated file `src/arrowExports.cpp`
*** > generated file `R/arrowExports.R`
*** Arrow C++ libraries found via pkg-config
PKG_CFLAGS=-DNDEBUG -I/usr/local/Cellar/apache-arrow/HEAD-81d3f26/include -DARROW_R_WITH_ARROW
PKG_LIBS=-L/usr/local/Cellar/apache-arrow/HEAD-81d3f26/lib -larrow_dataset -lparquet -larrow
clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DNDEBUG -I/usr/local/Cellar/apache-arrow/HEAD-81d3f26/include -DARROW_R_WITH_ARROW -I'/Users/romainfrancois/.R/library/4.0/cpp11/include' -I/usr/local/include   -fPIC  -Wall -O3 -Wall -Wimplicit-int-float-conversion -c array.cpp -o array.o
** libs
In file included from array.cpp:18:
In file included from ././arrow_types.h:22:
In file included from ././arrow_cpp11.h:37:
In file included from /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11.hpp:5:
In file included from /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/as.hpp:9:
/Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/protect.hpp:240:58: error: definition with same mangled name '_ZNK5cpp117protect8functionIFvPKczEEclIJRS3_EEEDTclclL_ZNSt3__17declvalIPS4_EEDTclsr3std3__1E9__declvalIT_ELi0EEEvEEspclsr3stdE7declvalIOT_EEEEDpSE_' as another definition
    decltype(std::declval<F*>()(std::declval<A&&>()...)) operator()(A&&... a) const {
                                                         ^
/Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/protect.hpp:240:58: note: previous definition is here
1 error generated.
make: *** [array.o] Error 1
ERROR: compilation failed for package ‘arrow’
* removing ‘/Users/romainfrancois/.R/library/4.0/arrow’
* restoring previous ‘/Users/romainfrancois/.R/library/4.0/arrow’

Exited with status 1.
```

File: cpp11test/src/test-protect.cpp [hunk 1]
```diff
@@ -37,6 +37,13 @@ context("unwind_protect-C++") {
     UNPROTECT(1);
   }
 
+  test_that("stop throws an unwind_exception") {
+    expect_error_as(cpp11::stop("error"), cpp11::unwind_exception);
+    expect_error_with(cpp11::stop("error"), "error");
```
Comment (romainfrancois) [inline cpp11test/src/test-protect.cpp:42]:
```suggestion
    expect_error(cpp11::stop("error"));
```

File: cpp11test/src/test-protect.cpp [hunk 2]
```diff
@@ -37,6 +37,13 @@ context("unwind_protect-C++") {
...
+    expect_error_as(cpp11::stop("error: %s", "message"), cpp11::unwind_exception);
+    expect_error_with(cpp11::stop("error: %s", "message"), "error: message");
```
Comment (romainfrancois) [inline cpp11test/src/test-protect.cpp:44]:
```suggestion
    expect_error(cpp11::stop("error: %s", "message"));
```

Comment (romainfrancois) [reply cpp11test/src/test-protect.cpp:44]: Looks like there is no `expect_error_with()`

File: cpp11test/src/test-protect.cpp (hunk 1 above)
Comment (bkietz) [reply cpp11test/src/test-protect.cpp:42]: `expect_error_with` checks the error message as well; IIUC `expect_error` just checks that it throws?

Comment (bkietz) [reply cpp11test/src/test-protect.cpp:42]: Never mind, I'll need to open a PR to testthat to add it

**Comment by jimhester** [comment]
If the code prior to https://github.com/r-lib/cpp11/pull/71 works on clang and the current code works on gcc one approach would be to ifdef them?

don't really enjoy playing compiler whack-a-mole though :(

**Comment by bkietz** [comment]
@jimhester if we can add a test with a minimal reproducer for the compilation failure I'd personally be comfortable adding the `#ifdef` and trying to golf it out in a follow up. @romainfrancois is working on one

**Comment by jimhester** [comment]
So is the ifdef solution the best option to go forward with this?

**Comment by bkietz** [comment]
I'm looking at Romain's reproducer. I'll either find a fix today or revert to #ifdefs

**Comment by bkietz** [comment]
@romainfrancois PTAL

**Comment by bkietz** [comment]
@romainfrancois it is internal, so we can rewrite as `safe_noreturn` anytime

**Pull request merged**